File: leetcode/144.binary-tree-preorder-traversal/cpp/preordertraversal_recur.cpp
#include <vector>

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    std::vector<int> preorderTraversal(TreeNode *root)
    {
        std::vector<int> res;
        traversal(root, res);
        return res;
    }

private:
    void traversal(TreeNode *root, std::vector<int> &res)
    {
        if (root == nullptr)
            return;

        res.push_back(root->val);
        traversal(root->left, res);
        traversal(root->right, res);
    }
};

File: leetcode/144.binary-tree-preorder-traversal/go/preordertraversal_recur.go
package preordertraversal

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func preorderTraversal(root *TreeNode) []int {
	res := make([]int, 0)

	return traversal(root, res)
}

func traversal(root *TreeNode, res []int) []int {
	if root == nil {
		return res
	}

	res = append(res, root.Val)
	res = traversal(root.Left, res)
	res = traversal(root.Right, res)

	return res
}
